File: Matrixes/gausPivoteoTotal.js
const mathjs = require('mathjs')

const totalPivot = (m, n, k) => {
    major = -1;
    posRow = -1;
    posColumn = -1;
    for (let i = k; i < n; i++) {
        for (let j = k; j < n; j++) {
            if (Math.abs(m[i][j]) > major) {
                major = Math.abs(m[i][j]);
                posRow = i;
                posColumn = j;
            }
        }
    }
    if(posRow !== k){
        const temp = m[k];
        m[k] = m[posRow];
        m[posRow] = temp;
    }
    if(posColumn !== k){
        m = mathjs.transpose(m);
        const temp2 = m[k];
        m[k] = m[posColumn];
        m[posColumn] = temp2;
        m = mathjs.transpose(m);
    }
    
    return { posColumn, matriz: m };
}


const generateMark = (n) => {
    const markResult = new Array(n)
    for (i = 0; i < n; i++){
        markResult[i] = i;
    } 
    return markResult;
}

const upperTriangular = (A, b, n) => {
    const stages = [];
    const marks = generateMark(n);
    m = mathjs.concat(A, b);
    stages.push([...m])
    for (let i = 0; i < n - 1; i++) {
        const { posColumn, matriz } = totalPivot(m, n, i);

        markAux = marks[i]
        marks[i] = marks[posColumn]
        marks[posColumn] = markAux;

        m = [...matriz];
        if (m[i][i] === 0) return { state: 'Error', message: 'A 0 was found on the diagonal' };
        for (let j = i + 1; j < n; j++) {
            if (m[j][i] !== 0) {
                m[j] = mathjs.subtract(mathjs.row(m, j), mathjs.multiply(mathjs.row(m, i), m[j][i] / m[i][i]))[0];
            }
        }
        stages.push([...m])
    }
    return { m, stages, marks };
}

const sustitution = (m, n) => {
    const results = new Array(n);
    results[n - 1] = m[n - 1][n] / m[n - 1][n - 1]

    for (let i = n - 2; i > -1; i--) {
        let summation = 0;
        for (let j = i + 1; j < n; j++) {
            summation += m[i][j] * results[j]
        }
        results[i] = (m[i][n] - summation) / m[i][i]
    }
    return results;
}


const orderResult = (x, marks) => {
    const result = new Array(x.length)
    marks.map((element,index)=>{
        result[element] = x[index]
    })
    return result;
}

module.exports = (A, b, n) => {
    const { m, stages, marks, state, message } = upperTriangular(A, b, n)

    if (state === 'Error') {
        return { state, message }
    }
    xprev = sustitution(m,n);
    x = orderResult(xprev,marks)

    return { state: 'Success', stages, x }
}

File: Matlab/gausPivoteoTotal.m
function [x] = gaussTotal(A,b,n)
    [m,marks] = upperTriangular(A,b,n);
    xPrev = sustitution(m,n);
    x = order(xPrev,marks);
end

function [x] = sustitution(m,n)
    x = zeros(1,n);
    x(n) = m(n,n+1)/m(n,n);
    for i = n-1:-1:1
        summation = 0;
        for j = i+1:1:n
            summation = summation + m(i,j)*x(j);
        end
        x(i) = (m(i,j+1) - summation)/m(i,i);
    end
end

function [m,posColumn] = totalPivot(m,n,k)
    major = -1;
    posRow = -1;
    posColumn = -1;

    for i = k:1:n
        for j = k:1:n
            if(abs(m(i,j))>major)
                major = abs(m(i,j));
                posRow = i;
                posColumn = j;
            end
        end
    end

    if(posRow ~= k)
        temp = m(k,:);
        m(k,:) = m(posRow,:);
        m(posRow,:) = temp;
    end

    if(posColumn ~= k)
        temp = m(:,k);
        m(:,k) = m(:,posColumn);
        m(:,posColumn) = temp;
    end
end

function [marks] = generateMark(n)
    marks = rand(1,n);
    for i = 1:1:n
        marks(i) = i;
    end
end

function [x] = order(xPrev,marks)
    x = rand(1,length(xPrev));
    for i=1:1:length(marks)
        x(marks(i)) = xPrev(i);
    end
end

function [m,marks] = upperTriangular(A,b,n)   
    m = [A b];
    [marks] = generateMark(n);
    disp('Etapa 0')
    disp(m)
    
    for i = 1:1:n-1
        [m,posColumn] = totalPivot(m,n,i);

        markAux = marks(i);
        marks(i) = marks(posColumn);
        marks(posColumn) = markAux;

        if (m(1,1)==0)
            disp('A 0 was found on the diagonal')
            return 
        end
        for j = i+1:1:n
            if(m(j,i)~=0)
                m(j,:) = m(j,:) - (m(j,i)/m(i,i)).*m(i,:);
            end
        end
        disp(['Etapa ',num2str(i)])
        disp(m)
    end
end